src/ipc.ts

```diff
@@ -0,0 +1,204 @@
+import { AsyncEventEmitter } from '@vladfrangu/async_event_emitter';
+import { Buffer } from 'node:buffer';
+import { randomUUID } from 'node:crypto';
+import { realpath } from 'node:fs/promises';
+import { createConnection, type Socket } from 'node:net';
+import process from 'node:process';
+import type { RPCClient } from './client.js';
+import { RPCMessagePayload } from './constants.js';
+
+enum OPCodes {
+	Handshake,
+	Frame,
+	Close,
+	Ping,
+	Pong,
+}
+
+async function getIPCPath(id: number) {
+	if (process.platform === 'win32') {
+		return `\\\\?\\pipe\\discord-ipc-${id}`;
+	}
+
+	const { XDG_RUNTIME_DIR, TMPDIR, TMP, TEMP } = process.env;
+	const prefix = await realpath(XDG_RUNTIME_DIR ?? TMPDIR ?? TMP ?? TEMP ?? '/tmp');
+
+	return `${prefix.replace(/\/$/, '')}/discord-ipc-${id}`;
+}
+
+async function getIPC(id = 0): Promise<Socket> {
+	const { promise, resolve, reject } = Promise.withResolvers<Socket>();
+
+	const path = await getIPCPath(id);
+
+	const onError = async () => {
+		if (id < 10) {
+			return getIPC(id + 1);
+		}
+
+		reject(new Error('Could not connect'));
+		return undefined;
+	};
+
+	const socket = createConnection(path, () => {
+		socket.removeListener('error', onError);
+
+		resolve(socket);
+	});
+
+	socket.once('error', onError);
+
+	return promise;
+}
+
+// async function findEndpoint(tries = 0): Promise<string> {
+// 	if (tries > 30) {
+// 		throw new Error('Could not find endpoint');
+// 	}
+
+// 	const endpoint = `http://127.0.0.1:${6_463 + (tries % 10)}`;
+
+// 	try {
+// 		const response = await fetch(endpoint);
+
+// 		if (response.status === 404) return endpoint;
+
+// 		return await findEndpoint(tries + 1);
+// 	} catch {
+// 		return findEndpoint(tries + 1);
+// 	}
+// }
+
+export function encode(op: number, data: RPCMessagePayload | string | {}) {
```

Comment:
`{}` here is redundant. Did you mean to type this as an empty object? If so, use `Record<string, never>`. `{}` represents any value except `null` and `undefined`.